test/e2e/persistent_volumes.go

```diff
@@ -18,159 +18,481 @@ package e2e
 
 import (
 	"encoding/json"
+	"fmt"
 	"time"
 
 	. "github.com/onsi/ginkgo"
 	"k8s.io/kubernetes/pkg/api"
+	apierrs "k8s.io/kubernetes/pkg/api/errors"
 	"k8s.io/kubernetes/pkg/api/resource"
 	"k8s.io/kubernetes/pkg/api/testapi"
 	"k8s.io/kubernetes/pkg/api/unversioned"
 	client "k8s.io/kubernetes/pkg/client/unversioned"
 	"k8s.io/kubernetes/test/e2e/framework"
 )
 
-// Clean both server and client pods.
-func persistentVolumeTestCleanup(client *client.Client, config VolumeTestConfig) {
+// Delete the nfs-server pod.
+func nfsServerPodCleanup(c *client.Client, config VolumeTestConfig) {
 	defer GinkgoRecover()
 
-	podClient := client.Pods(config.namespace)
+	podClient := c.Pods(config.namespace)
 
 	if config.serverImage != "" {
-		err := podClient.Delete(config.prefix+"-server", nil)
+		podName := config.prefix + "-server"
+		err := podClient.Delete(podName, nil)
 		if err != nil {
-			framework.Failf("Failed to delete the server pod: %v", err)
+			framework.Logf("Delete of %v pod failed: %v", podName, err)
 		}
 	}
 }
 
-func deletePersistentVolume(c *client.Client, pv *api.PersistentVolume) {
-	// Delete the PersistentVolume
-	framework.Logf("Deleting PersistentVolume")
+// Delete the PV. Fail test if delete fails.
+func deletePersistentVolume(c *client.Client, pv *api.PersistentVolume) (*api.PersistentVolume, error) {
+
+	if pv == nil {
+		return nil, fmt.Errorf("PV to be deleted is nil")
+	}
+
+	By("Deleting PersistentVolume")
+
+	framework.Logf("Deleting PersistentVolume %v", pv.Name)
 	err := c.PersistentVolumes().Delete(pv.Name)
 	if err != nil {
-		framework.Failf("Delete PersistentVolume failed: %v", err)
+		return pv, fmt.Errorf("Delete() PersistentVolume %v failed: %v", pv.Name, err)
+	}
+
+	// Wait for PersistentVolume to delete
+	framework.WaitForPersistentVolumeDeleted(c, pv.Name, 3*time.Second, 90*time.Second)
+
+	// Double check that the PV is really deleted. It is important to not
+	// delete the nfs-server pod until after the pv is deleted/recycled.
+	pv, err = c.PersistentVolumes().Get(pv.Name)
```

Comment:
I don't think this check is necessary, it does the same thing as `WaitForPersistentVolumeDeleted` called above. Or did you experence `WaitForPersistentVolumeDeleted` returning no error when a PV still existed?